syft/pkg/catalog.go

```diff
@@ -41,43 +39,133 @@ func (c *Catalog) PackageCount() int {
 
 // Package returns the package with the given ID.
 func (c *Catalog) Package(id ID) *Package {
-	return c.byID[id]
+	v, exists := c.byID[id]
+	if !exists {
+		return nil
+	}
+	return v
+}
+
+// PackagesByPath returns all packages that were discovered from the given path.
+func (c *Catalog) PackagesByPath(path string) []*Package {
+	return c.Packages(c.idsByPath[path])
 }
 
-// PackagesByFile returns all packages that were discovered from the given source file reference.
-func (c *Catalog) PackagesByFile(location source.Location) []*Package {
-	return c.byFile[location]
+// Packages returns all packages for the given ID.
+func (c *Catalog) Packages(ids []ID) (result []*Package) {
+	for _, i := range ids {
+		p, exists := c.byID[i]
+		if exists {
+			result = append(result, p)
+		}
+	}
+	return result
 }
 
 // Add a package to the Catalog.
 func (c *Catalog) Add(p Package) {
-	if p.id != 0 {
-		log.Errorf("package already added to catalog: %s", p)
-		return
-	}
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	p.id = ID(nextPackageID)
-	nextPackageID++
+	_, exists := c.byID[p.ID]
+	if exists {
+		log.Errorf("package ID already exists in the catalog : id=%+v %+v", p.ID, p)
+		return
+	}
+
+	if p.ID == "" {
+		p.ID = newID()
```

Comment:
This is a bit concerning — the nondeterminism — especially since this gets exposed outside of the implementation. But the same could be said of any persistence/store-oriented object like a database table. No change needed right now — just noting a non-ideal trait. Open to further discussion

Reply:
IDs by nature should be nondeterministic, we shouldn't be depending on a particular value and claim this is a ID --so the change from an incrementing number (which is still not deterministic since package order may vary on addition depending on the cataloger implementation... if we add concurrency for instance).

As you noted, this is a impactful change though since this is being exposed (where previously it was not). You couldn't reference theses IDs across syft runs, since the IDs are random, however, we could in the future provide a stable/deterministic package "fingerprint" field derived from the  package property values.

Reply:
This issue I think is relevant wrt to IDs https://github.com/anchore/syft/issues/166 - definitely something that has come up before

Reply:
Ah yeah, I remember that one. I don't know about "should be nondeterministic" — I think the important attribute of IDs is uniqueness. Something consumers can use to positively find a single item.

I'm okay with this code going in as is, I was more thinking of what @alfredodeza mentioned — this has come up before, and it warrants caution.

Reply:
@alfredodeza good reference, adding to the PR to close it; this PR takes your suggestion to allow IDs to be an exported field on the `Package` struct and loosens assumptions about ID with regard to catalog assignment.

@luhring Agreed that IDs should be unique, that's not in question. 

I want to clarify why I think non-determinisim is good quality to have in an ID system as a rule of thumb (note, not as a strict "always-do-this" rule).

In #166 , the fact that the ID was being derived from a global value is bad, but what is also bad is that tests were depending on predictable ID values for correctness, and the value was not being controlled by the test (which is what @alfredodeza  was looking to do). If the ID was random from the start such accidental dependence on ID values would not be possible --the tests would not have passed to begin with.

Exposing the existing IDs with the current approach (incrementing number) an end user may assume that there is an implied order to packages. They may assume that the order is stable between runs --which is true, but is only a coincidence due to the current implementation. The truth is, the ID should not convey any intrinsic information --by design or accidentally-- other than the fact that your guaranteed to not have duplicates.

Anywho, I know you weren't asking for a change here @luhring  but wanted to clarify my statements on IDs.